rhombus/private/bytes.rkt

```diff
@@ -25,79 +29,109 @@
 
-(define-name-root Bytes
-  #:fields
-  ([make make-bytes]
-   [length bytes-length]
-   [subbytes sub_bytes]
-   copy
-   copy_from
-   utf8_string
-   latin1_string
-   locale_string))
+(define (check-bytes who bs)
+  (unless (bytes? bs)
+    (raise-argument-error* who rhombus-realm "Bytes" bs)))
+
+(define (check-nonneg-int who n)
+  (unless (exact-nonnegative-integer? n)
+    (raise-argument-error* who rhombus-realm "NonnegInt" n)))
+
+(define/arity (Bytes.get b i)
+  (check-bytes who b)
+  (check-nonneg-int who i)
+  (bytes-ref b i))
 
-(define/arity #:name subbytes (sub_bytes str [start 0] [end (and (bytes? str) (bytes-length str))])
+(define/arity (Bytes.set b i x)
+  (check-bytes who b)
+  (check-nonneg-int who i)
+  (bytes-set! b i x))
+
+(define/arity (Bytes.append b1 b2)
+  (check-bytes who b1)
+  (check-bytes who b2)
+  (bytes-append b1 b2))
+
+(define/arity (Bytes.make len [val 0])
   #:static-infos ((#%call-result #,bytes-static-infos))
-  (subbytes str start end))
-
-(define/arity (utf8_string s [err-char #f] [start 0] [end (and (bytes? s) (bytes-length s))])
-  #:static-infos ((#%call-result #,indirect-string-static-infos))
-  (string->immutable-string (bytes->string/utf-8 s err-char start end)))
-(define/arity (latin1_string s [err-char #f] [start 0] [end (and (bytes? s) (bytes-length s))])
-  #:static-infos ((#%call-result #,indirect-string-static-infos))
-  (string->immutable-string (bytes->string/latin-1 s err-char start end)))
-(define/arity (locale_string s [err-char #f] [start 0] [end (and (bytes? s) (bytes-length s))])
-  #:static-infos ((#%call-result #,indirect-string-static-infos))
-  (string->immutable-string (bytes->string/locale s err-char start end)))
-(define/arity (copy s)
+  (check-nonneg-int who len)
+  (make-bytes len val))
+
+(define/method (Bytes.length bstr)
+  (check-bytes who bstr)
+  (bytes-length bstr))
+
+(define/method (Bytes.subbytes bstr [start 0] [end (and (bytes? bstr) (bytes-length bstr))])
   #:static-infos ((#%call-result #,bytes-static-infos))
-  (bytes-copy s))
-(define/arity (copy_from s dest-start src [src-start 0] [src-end (and (bytes? src) (bytes-length src))])
+  (check-bytes who bstr)
+  (check-nonneg-int who start)
+  (check-nonneg-int who end)
+  (subbytes bstr start end))
+
+(define (check-maybe-char who c)
+  (unless (or (not c) (char? c))
+    (raise-argument-error who rhombus-realm "maybe(Char)" c)))
+
+(define-syntax (define-string stx)
+  (syntax-parse stx
+    [(_ utf8 utf-8)
+     #:with method-name (datum->syntax #'utf8 (string->symbol (format "Bytes.~a_string" (syntax-e #'utf8))))
+     #:with fn-name (datum->syntax #'utf-8 (string->symbol (format "bytes->string/~a" (syntax-e #'utf-8))))
+     #'(define/method (method-name bs [err-char #f] [start 0] [end (and (bytes? bs) (bytes-length bs))])
+         #:static-infos ((#%call-result #,indirect-string-static-infos))
+         (check-maybe-char who err-char)
```

Comment:
This is a case where functions like `string->bytes/utf-8` cannot be used directly, but maybe the checking built into those methods should be relied on, instead of checking again here. Note that the checks here are incomplete, because there are constraints on `start` versus `end` versus the length of the string.